### DISCUSSION TITLE
Is there a way to disable (parts of) the sysroot cache?

### DISCUSSION THREAD

**Discussion by icculus** [emscripten-core/emscripten#17647] [General]
(This felt like it shouldn't be a bug report, but let me know if it should be instead!)

This is with 3.1.18 (just a fresh emsdk clone and `emsdk install latest` from today).

I'm trying to build SDL from source, but Emscripten is supplying its own copy of SDL in 
emsdk/upstream/emscripten/cache/sysroot, which is causing me issues like this, where it is favoring the sysroot headers over the ones included with the project _while building the project_:

```
[1/122] Building C object CMakeFiles/SDL2-static.dir/src/SDL.c.o
FAILED: CMakeFiles/SDL2-static.dir/src/SDL.c.o 
/home/icculus/projects/emsdk/upstream/emscripten/emcc -DSDL_STATIC_LIB -DUSING_GENERATED_CONFIG_H -I/home/icculus/projects/SDL/cmake-build/include -I/home/icculus/projects/SDL/cmake-build/include-config-debug -idirafter "/home/icculus/projects/SDL/src/video/khronos"  -Wall -fno-strict-aliasing -Werror=declaration-after-statement -Wdeclaration-after-statement -fvisibility=hidden -Wshadow -fdiagnostics-color=always  -DSDL_BUILD_MAJOR_VERSION=2 -DSDL_BUILD_MINOR_VERSION=23 -DSDL_BUILD_MICRO_VERSION=2 -g -Wno-warn-absolute-paths -MD -MT CMakeFiles/SDL2-static.dir/src/SDL.c.o -MF CMakeFiles/SDL2-static.dir/src/SDL.c.o.d -o CMakeFiles/SDL2-static.dir/src/SDL.c.o -c /home/icculus/projects/SDL/src/SDL.c
In file included from /home/icculus/projects/SDL/src/SDL.c:47:
In file included from /home/icculus/projects/SDL/cmake-build/include/SDL_bits.h:31:
/home/icculus/projects/SDL/cmake-build/include/SDL_stdinc.h:186:5: warning: declaration shadows a variable in the global scope [-Wshadow]
    SDL_FALSE = 0,
    ^
/home/icculus/projects/emsdk/upstream/emscripten/cache/sysroot/include/SDL/SDL_stdinc.h:117:5: note: previous declaration is here
    SDL_FALSE = 0,
    ^
```

This isn't an issue on our GitHub Actions builds, because we're building with a significantly older emsdk at the moment, but this will eventually cause up problems when we upgrade.

Is there some way to disable the cache, either entirely or just parts of it, temporarily or permanently?

Thanks!
--ryan.

**Comment by kripken** [comment]
Hmm, I'd expect that to just work since include paths you include should have precedence over system ones... so that is odd. Somehow your SDL2 build ends up including headers from SDL1 in the sysroot.

Perhaps @sbc100 has an idea? Otherwise, maybe seeing the verbose output of emcc/clang (`-v`) might help to see what's going on there.

In general, you can clear the sysroot cache (`emcc --clear-cache`) and then populate it only with what you need. If you don't specify `-sUSE_SDL=2` then SDL2 won't be built. But the problem here is with SDL1 which is present by default in the sysroot, so I can't think of a simple workaround here.

**Reply by sbc100** [reply]
Hmm... its possible I broke this with #17475.   User include paths should always come first so this is certainly a bug. 

The fix I would recommend right now is to use `-sUSE_SDL=0`.  Emscripten, for better or worse, defaults to `-sUSE_SDL=1` which is injecting those paths.  That way it won't matter about the order because we won't be injecting those extra headers at all.

**Reply by sbc100** [reply]
Please let me know if `-sUSE_SDL=0` fixes your issue.  If it does I think that is probably the correct fix.  

You could also build with `-sSTRICT` which turns of a lot of legacy/questionable features, including setting `-sUSE_SDL=1` by default, so building in this mode would ideal.

**Comment by icculus** [comment]
> Please let me know if -sUSE_SDL=0 fixes your issue. If it does I think that is probably the correct fix.

It does! I've stuck a fix in the build script here for now: https://github.com/libsdl-org/SDL/commit/51be30f3cd29682cd94fcbfbe18eeac492095a05

Thanks!
